libp2p/autotls/manager.nim

```diff
@@ -0,0 +1,292 @@
+    raise newException(AutoTLSError, "Cannot issue new certificate: peerInfo not set")
+
+  # generate autotls domain string: "*.{peerID}.libp2p.direct"
+  let base36PeerId = encodePeerId(peerInfo.peerId)
+  let baseDomain = api.Domain(base36PeerId & "." & AutoTLSDNSServer)
+  let domain = api.Domain("*." & baseDomain)
+
+  let acmeClient = self.acmeClient.valueOr:
+    raise newException(AutoTLSError, "Cannot find ACMEClient on manager")
+
+  trace "Requesting ACME challenge"
+  let dns01Challenge = await acmeClient.getChallenge(@[domain])
+  let keyAuth = acmeClient.genKeyAuthorization(dns01Challenge.dns01.token)
+  let strMultiaddresses: seq[string] = peerInfo.addrs.mapIt($it)
+  let payload = %*{"value": keyAuth, "addresses": strMultiaddresses}
+  let registrationURL = parseUri("https://" & AutoTLSBroker & "/v1/_acme-challenge")
+
+  trace "Sending challenge to AutoTLS broker"
+  let (bearer, response) =
+    await self.brokerClient.send(registrationURL, peerInfo, payload, self.bearer)
+  if self.bearer.isNone:
+    # save bearer token for future
+    self.bearer = Opt.some(bearer)
+  if response.status != HttpOk:
+    raise newException(
+      AutoTLSError, "Failed to authenticate with AutoTLS Broker at " & AutoTLSBroker
+    )
+
+  debug "Waiting for DNS record to be set"
+  let ipAddress = self.ipAddress.valueOr:
+    raise newException(AutoTLSError, "No public IPv4 address specified")
+
+  let dnsSet = await self.checkDNSRecords(ipAddress, baseDomain, keyAuth)
+  if not dnsSet:
+    raise newException(AutoTLSError, "DNS records not set")
+
+  debug "Notifying challenge completion to ACME and downloading cert"
+  let certResponse = await acmeClient.getCertificate(domain, dns01Challenge)
+
+  trace "Installing certificate"
+  try:
+    self.cert = Opt.some(TLSCertificate.init(certResponse.rawCertificate))
+    self.certExpiry = Opt.some(asMoment(certResponse.certificateExpiry))
+  except TLSStreamProtocolError:
+    raise newException(AutoTLSError, "Could not parse downloaded certificates")
+  self.certReady.fire()
+
+proc manageCertificate(
+    self: AutoTLSManager
+): Future[void] {.async: (raises: [AutoTLSError, ACMEError, CancelledError]).} =
+  trace "Starting AutoTLS manager"
+  let ipAddress = self.ipAddress.valueOr:
+    try:
+      let ip = self.ipAddress.valueOr:
+        checkedGetPrimaryIPAddr()
+      if not ip.isIPv4:
+        raise newException(AutoTLSError, "Host does not have an IPv4 address")
+      if not ip.isPublic:
+        raise newException(AutoTLSError, "Host does not have a public IPv4 address")
+      ip
+    except AutoTLSError as exc:
+      raise exc
+    except CatchableError as exc:
+      raise newException(
+        AutoTLSError, "Unexpected error while getting primary IP address for host", exc
+      )
+  self.ipAddress = Opt.some(ipAddress)
+
+  debug "Registering ACME account"
+  if self.acmeClient.isNone:
+    self.acmeClient = Opt.some(await ACMEClient.new(acmeServerURL = self.acmeServerURL))
```

Comment:
why not initialize this in new?

Reply:
okay got it. 

if you don't like `AutoTLSManager.new` to be async why is it okay for `ACMEClient.new` ?

Reply:
> you don't like AutoTLSManager.new to be async

I don't mind either way, but if `AutoTLSManager` were to be `async`, then so would `withAutotls`, and that breaks the switch building process. I see your point, but changing ACMEClient's conventions is not in the scope of this PR.